src/torcontrol.cpp

```diff
@@ -373,9 +373,18 @@ void TorController::auth_cb(TorControlConnection& _conn, const TorControlReply&
     if (reply.code == 250) {
         LogPrint(BCLog::TOR, "tor: Authentication successful\n");
 
-        // Now that we know Tor is running setup the proxy for onion addresses
-        // if -onion isn't set to something else.
-        if (gArgs.GetArg("-onion", "") == "") {
+        // Now that we know Tor is running, possibly set the proxy for onion addresses.
+        bool allow_outbound_onion{true};
+        if (gArgs.IsArgSet("-onlynet")) {
+            allow_outbound_onion = false;
+            for (const auto& net : gArgs.GetArgs("-onlynet")) {
+                if (ParseNetwork(net) == NET_ONION) {
+                    allow_outbound_onion = true;
+                    break;
+                }
+            }
+        }
+        if (gArgs.GetArg("-onion", "") == "" && allow_outbound_onion) {
```

Comment:
A function seems nicer than a localvar here (optionally declare in header file for unit testing)
```diff
+bool AllowOutboundOnion()
+{
+    if (gArgs.IsArgSet("-onlynet")) return false;
+    for (const auto& net : gArgs.GetArgs("-onlynet")) {
+        if (ParseNetwork(net) == NET_ONION) return true;
+    return false;
+}
+
 void TorController::auth_cb(TorControlConnection& _conn, const TorControlReply& reply)
 {
     if (reply.code == 250) {
         LogPrint(BCLog::TOR, "tor: Authentication successful\n");
 
         // Now that we know Tor is running, possibly set the proxy for onion addresses.
-        bool allow_outbound_onion{true};
-        if (gArgs.IsArgSet("-onlynet")) {
-            allow_outbound_onion = false;
-            for (const auto& net : gArgs.GetArgs("-onlynet")) {
-                if (ParseNetwork(net) == NET_ONION) {
-                    allow_outbound_onion = true;
-                    break;
-                }
-            }
-        }
-        if (gArgs.GetArg("-onion", "") == "" && allow_outbound_onion) {
+        if (gArgs.GetArg("-onion", "").empty() && AllowOutboundOnion()) {
             CService resolved(LookupNumeric("127.0.0.1", 9050));
             proxyType addrOnion = proxyType(resolved, true);
```

Reply:
Tested with this patch on mainnet:
- `onlynet=i2p`: only I2P outbound connections are made (i2pd 2.23)
- `onlynet=ipv4`: only IPv4 outbound connections are made
- `onlynet=onion`: only onion outbound connections are made (tor 0.4.6)

Reply:
Tested combinations (all tests using the proposed function version ;)
- `onlynet=i2p` and `onlynet=onion`: only I2P and onion outbound connections are made
- `onlynet=ipv4` and `onlynet=onion`: only IPv4 and onion outbound connections are made

(Test coverage for these cases--single onlynet and combinations--would be great.)

Reply:
> Test coverage for these cases--single onlynet and combinations--would be great

Agree

Reply:
I think it is fine either way. A function would make more sense if this is used in more than one place. The code is now covered by a new functional test `feature_onlynet.py`. Left it as is.

Reply:
I think the function is not only shorter but also clearer and easier to reason about.